### ISSUE TITLE
many_many filters no longer work on 3.0

### ISSUE THREAD

**Issue by silverstripe-elliot** [silverstripe/silverstripe-fulltextsearch#205]
I'm trying to index a many_many field so that I can filter on it. This used to work in 2.3.5 (or thereabouts) but no longer works in 3.0

```MyIndex.php
public function init() {
   $this->addClass(MyPage::class);
...
   $this->addFilterField('Categories.ID', 'Int');
...
}
```

```MyPage.php
...
    private static $many_many = [
      'Categories' => TaxonomyTerm::class,
   ]
...
```

The schema gets generated correctly: `<field name='MyPage_Categories_ID' type='tint' indexed='true' stored='true' multiValued='true'/>` but the data never appears on the record in Solr. This means it's impossible to filter on it at query time.

One of my colleagues has experienced a similar problem, and found that one resolution is to rollback to 2.3.5. Unfortunately, my project is on SS4 and I don't have this option.

**Comment by NightJar** [comment]
Hi @silverstripe-elliot - since you've worked a wee bit with this here module before, can you think of any valid reason someone would look up this kind of information on the requested class _only_?
i.e. the logic that adds such a relationship to your index now uses `Config::UNINHERITED`, which I'm assuming could be your issue (i.e. `'Categories'` is defined on a parent class?)

Compare now:
https://github.com/silverstripe/silverstripe-fulltextsearch/blob/master/src/Search/Indexes/SearchIndex.php#L160

With then:
https://github.com/silverstripe/silverstripe-fulltextsearch/blob/2.3/code/search/SearchIndex.php#L99

**Comment by silverstripe-elliot** [comment]
Sorry, I'm not sure I understand the question. I noticed that Github ate my edits a little bit so I'll try to clarify.

The many_many relation exists on a ResourcePage, which extends Page. The ResourcePage has a many_many relationship with Categories, which are managed with the TaxonomyTerm module. It's not defined on the parent. 

```
class ResourcePage extends Page {
    private static $many_many = [
        'Categories' => TaxonomyTerm::class
    ];
}
```

**Comment by robbieaverill** [comment]
Re: @NightJar's findings, it's correct to note that there's a change in behaviour (whether intentional or not), in that the SS3 version pulls an _inherited_ config prop of many_many whereas in SS4 it's _uninherited_ (reference: SS3 `DataObject::manyMany`, SS4 `SearchIndex:: fieldData`).

This change in behaviour occurred in https://github.com/silverstripe/silverstripe-fulltextsearch/commit/2b3357565e302552d32e512395c019ff15b8f9e9, and is released under 2.4.0 as well as the 3.x line for SS3.

@mfendeksilverstripe would you be able to provide some insight into the reasons for the logic change in that commit?

**Comment by NightJar** [comment]
It is interesting to note though that Elliot's issue occurs where the relationship _is_ defined on the class being added to the index. This results in it being in the schema, where as the code implies (I didn't follow it _that_ deep) that if it were _inherited_ then it wouldn't feature in the schema at all.

**Comment by robbieaverill** [comment]
Worth revalidating after the latest merge

**Comment by NightJar** [comment]
Your variant application detection & query business?

**Comment by elliot-sawyer** [comment]
My own research on this matched @NightJar 's findings. If you have a look in the schema.xml for your Solr index after running Solr_Configure, you'll find that the schema `MyPage_Categories_ID` is correctly detected. It's even loaded into Solr if you check it via the admin console. However, that field `MyPage_Categories_ID` never appears on any record in Solr itself when you query for `*`. That implies it's never being scanned and/or committed to the Solr server. 

I didn't look into any of the variants possibly misbehaving, but I believe the issue occurs somewhere during re-indexing.

**Comment by NightJar** [comment]
I guess the first place to start here @elliot-sawyer is to ask which modules you have installed (i.e. which set of `SearchVariant` subclasses are applying to your site the exhibits the issue). I'm assuming a basic set of those which come out of the box, so to speak - but if we need to consider the likes of e.g. `fluent` also that would be handy to know :)

**Comment by elliot-sawyer** [comment]
The only SearchVariant that I'm aware of is the one that excludes draft content (SearchVariantVersioned, I think?). There's nothing like Subsites, Translatable, or Fluent being used.

**Comment by mfendeksilverstripe** [comment]
@robbieaverill 

I don't think my changeset in related to this issue, but then again it was almost a year ago.

The `Config::UNINHERITED` is used because I introduced extra iteration loops by removing the `continue 2;` statement. If you trace the building of the index you will see that my changeset actually increases the number of iterations because in the original version some indexes were missed. The `Config::UNINHERITED` is needed because we are iterating through all classes via the hierarchy so we are bound to find the class that has the relation defined.

A couple of unit tests were added to cover these change set as well so it might be a good idea to look at it. I also did some diffs on the `index` xml file before and after my change set. If you want to dig deeper into this issue I recommend comparing the index xml files between SOLR versions.
